### PR TITLE
Fix user auth routes, waiting room dialogue/socket, and run code from BE PR consolidation merge.

### PR DESCRIPTION
**Pull request by kyi193**
Hey guys, I went ahead of merged all the current PRs together into this branch. It will include:

- Tim: Create room and waiting room dialogue + interview model

- Nafis: Execute code and return output from the backend

- Kevin Yi: Fix user auth routing + waiting room sockets

Some things to note: 

1. We won't be needing cors anymore as we're calling directly from the backend server now (no need to call from localhost:3000 anymore)
2. We'll be using api functions for all backend calls on the front end (located in utils/apiFunctions.js)
3. We'll create a global socket instance on app load that is created in context. It can be accessed inside of state.socket

Might be missing a few other points but those were the most important ones to keep in mind.

### CODE REVIEW

File: client/src/context/store.js
Comment (skassam21) [inline client/src/context/store.js:13]: Should this go into a useEffect? Has this been working fine for you?

File: client/src/context/store.js
Comment (skassam21) [inline client/src/context/store.js:35]: beacuse you return ...state, do you need to keep passing it through?

**Review by skassam21 — COMMENTED** [review]
Can we merge this after we merge Tim's current PR, there is a bunch of overlap

File: client/src/context/store.js
Comment (kyi193) [reply client/src/context/store.js:13]: Yeah this has been working perfectly fine without being inside a useEffect

File: client/src/context/store.js
Comment (kyi193) [reply client/src/context/store.js:35]: I thought so too but for some reason state.socket was returning undefined in Lobby.js unless I explicitly added it in like this.

**Comment by kyi193** [comment]
I fixed 2 bugs this morning:

1. A user exiting an already full lobby would disconnect another use already in the lobby
2. Starting an interview would cause people who joined the lobby and then quit to still redirect them to the interview page

**Review by skassam21 — APPROVED** [review]

**Pull request merged**
